File: benchmarks/values/closure_iter.rb
t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)

n = 2_000_000

sum = (0...n)
  .map { |x| x * 3 + 1 }
  .select { |x| x % 2 == 0 }
  .map { |x| x / 2 }
  .reduce(0, :+)

t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
puts sum
puts "elapsed: #{'%.3f' % (t1 - t0)}s"

File: benchmarks/values/closure_iter.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define N 2000000

int main(void) {
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    /* map: x -> x * 3 + 1, filter: x % 2 == 0, map: x -> x / 2, reduce: sum */
    long long sum = 0;
    for (int x = 0; x < N; x++) {
        long long v = (long long)x * 3 + 1;
        if (v % 2 == 0) {
            sum += v / 2;
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) / 1e9;
    printf("%lld\n", sum);
    printf("elapsed: %.3fs\n", elapsed);
    return 0;
}
